Add User Registration Functionality

ShastkivRuslan opened the pull request:
This pull request introduces user registration functionality into the system, including the new User entity, corresponding DTOs, an authentication controller, and an error handling mechanism. Key changes include:

Addition of User Entity:

Fields:

- id (Long, PK)
- email (String, not null, unique)
- password (String, not null)
- firstName (String, not null)
- lastName (String, not null)
- shippingAddress (String)

Creation of DTO Classes:

- UserRegistrationRequestDto: for receiving data about the new user during registration.
- UserResponseDto: for returning information about the registered user.

Addition of Mapper:

- Utilizing MapStruct for transforming between the User entity and the corresponding DTOs.

AuthenticationController:

- Added method register(UserRegistrationRequestDto request) throws RegistrationException, which handles user registration requests.

Data Validation:

- Implementing validation annotations in the new DTO classes to ensure the correctness of incoming data.
- Added the @PasswordMatcher annotation to check that the password and repeatPassword fields in UserRegistrationRequestDto are the same.

Error Handling:

- Implemented error handling for registration using GlobalExceptionHandler for centralized management of exceptions that may occur during registration.

ShastkivRuslan committed:
added User model;

ShastkivRuslan committed:
added liquibase script for creating users table;

ShastkivRuslan committed:
changes in liquibase script:
->changed length for password row from 30 to 255 characters(for safe saving hashed passwords)
-fixed typo in shipping address

ShastkivRuslan committed:
-replaced bokdto`s to book package
-created user package
-created UserRegistrationRequestDto

ShastkivRuslan committed:
-created UserDto;

ShastkivRuslan committed:
-added UserMapper with 2 methods;

ShastkivRuslan committed:
mvn clean package -> passed;

ShastkivRuslan committed:
Implement user authentication and registration functionality

- Added AuthenticationController to handle login requests.
- Introduced RegistrationException for better error handling during user registration.
- Created UserService and UserServiceImpl for user management.
- Updated BookController, BookService, and BookServiceImpl to reflect recent changes.
- Added UserRepository for user data persistence.

ShastkivRuslan committed:
Add @PasswordMatcher annotation for password and repeatPassword validation in UserRegistrationRequestDto

ShastkivRuslan committed:
-added swagger documentation;

ShastkivRuslan committed:
-bugfix;

ShastkivRuslan committed:
mvn clean package -> done;

ShastkivRuslan committed:
some improvements;

ShastkivRuslan committed:
fix build;

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/exception/GlobalExceptionHandler.java:
```diff
@@ -34,4 +36,13 @@ public ResponseEntity<Map<String, String>> handleEntityNotFoundException(
+        Map<String, String> errorResponse = new HashMap<>();
+        errorResponse.put("error", exception.getMessage());
+
+        return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
```
```suggestion
        return new ResponseEntity<>(errorResponse, HttpStatus.CONFLICT);
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public UserDto register(UserRegistrationRequestDto registrationRequestDto) {
+        if (userRepository.findByEmail(registrationRequestDto.email()).isPresent()) {
```
```suggestion
        if (userRepository.existsByEmail(registrationRequestDto.email())) {
```

Elena-Bruyako commented on src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+                    + registrationRequestDto.email() + "] already registered!"
+            );
+        }
+        return userMapper.toDto(
```
don't call several methods inline, use local variables

Elena-Bruyako reviewed (changes requested):
see comments

ShastkivRuslan committed:
Changes after mentor review:
- Updated GlobalExceptionHandler to return HttpStatus.CONFLICT for duplicate registration attempts instead of HttpStatus.BAD_REQUEST.
- Refactored UserServiceImpl:
- - Changed the method for checking if an email already exists to userRepository.existsByEmail(registrationRequestDto.email()) for improved readability and efficiency.
- - Introduced local variables instead of calling multiple methods inline, enhancing code clarity.

liliia-ponomarenko commented on src/main/java/ruslan/shastkiv/bookstore/exception/GlobalExceptionHandler.java:
```diff
@@ -34,4 +36,13 @@ public ResponseEntity<Map<String, String>> handleEntityNotFoundException(
+
+        return new ResponseEntity<>(errorResponse, HttpStatus.CONFLICT);
+    }
+
```
```suggestion
```

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,39 @@
+                    nullable: false
+              - column:
+                  name: email
+                  type: varchar(50)
```
Better use the same length validators with dto. Otherwise, you may get an exception when saving to the DB.

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,39 @@
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: first_name
+                  type: varchar(30)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: last_name
+                  type: varchar(30)
+                  constraints:
+                    nullable: false
```
the same

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/02-create-users-table.yaml:
```diff
@@ -0,0 +1,39 @@
+                    nullable: false
+              - column:
+                  name: shipping_address
+                  type: varchar(255)
```
leave one empty line at the end of the file (check other files)

liliia-ponomarenko reviewed (approved):
Good job! But please fix my comments before merging ;)

ShastkivRuslan committed:
- Update `GlobalExceptionHandler.java` to align with feedback.
- Adjust varchar length in `02-create-users-table.yaml` to match DTO validators for consistency.
- Add missing empty line at the end of `02-create-users-table.yaml` for standard formatting.
